iceoryx_utils/test/moduletests/test_posix_timer.cpp

```diff
@@ -406,3 +407,117 @@ TEST_F(Timer_test, GetOverrunsFailsWithNoCallback)
+    std::atomic_bool hasTerminated{false};
+    auto errorHandlerGuard = iox::ErrorHandler::SetTemporaryErrorHandler(
+        [&](const iox::Error, const std::function<void()>, const iox::ErrorLevel) { hasTerminated = true; });
+
+    Timer sut(TIMEOUT,
+              [] { std::this_thread::sleep_for(std::chrono::milliseconds(TIMEOUT.milliSeconds<int>() * 10)); });
+
+    sut.start(Timer::RunMode::PERIODIC, Timer::CatchUpPolicy::IMMEDIATE);
+
+    std::this_thread::sleep_for(std::chrono::milliseconds(TIMEOUT.milliSeconds<int>() * 10));
+    TIMING_TEST_EXPECT_FALSE(hasTerminated);
+});
+
+TIMING_TEST_F(Timer_test, HardTimerTerminatesWhenCallbackIsLongerThenTriggerTime, Repeat(5), [&] {
+    std::atomic_bool hasTerminated{false};
+    auto errorHandlerGuard = iox::ErrorHandler::SetTemporaryErrorHandler(
+        [&](const iox::Error, const std::function<void()>, const iox::ErrorLevel) { hasTerminated = true; });
+
+    Timer sut(TIMEOUT,
+              [] { std::this_thread::sleep_for(std::chrono::milliseconds(TIMEOUT.milliSeconds<int>() * 10)); });
+
+    sut.start(Timer::RunMode::PERIODIC, Timer::CatchUpPolicy::TERMINATE);
+
+    std::this_thread::sleep_for(std::chrono::milliseconds(TIMEOUT.milliSeconds<int>() * 10));
+    TIMING_TEST_EXPECT_TRUE(hasTerminated);
+});
+
+TIMING_TEST_F(Timer_test, SoftToHardTimerChangesBehaviorToTerminate, Repeat(5), [&] {
+    std::atomic_bool hasTerminated{false};
+    auto errorHandlerGuard = iox::ErrorHandler::SetTemporaryErrorHandler(
+        [&](const iox::Error, const std::function<void()>, const iox::ErrorLevel) { hasTerminated = true; });
+
+    Timer sut(TIMEOUT,
+              [] { std::this_thread::sleep_for(std::chrono::milliseconds(TIMEOUT.milliSeconds<int>() * 10)); });
+
+    sut.start(Timer::RunMode::PERIODIC, Timer::CatchUpPolicy::SKIP_TO_NEXT_BEAT);
+    std::this_thread::sleep_for(std::chrono::milliseconds(TIMEOUT.milliSeconds<int>() * 10));
+    sut.restart(TIMEOUT, Timer::RunMode::PERIODIC, Timer::CatchUpPolicy::TERMINATE);
+    std::this_thread::sleep_for(std::chrono::milliseconds(TIMEOUT.milliSeconds<int>() * 10));
+
+    TIMING_TEST_EXPECT_TRUE(hasTerminated);
+});
+
+TIMING_TEST_F(Timer_test, SoftTimerSkipsCallbackWhenStillRunning, Repeat(5), [&] {
+    std::atomic_int counter{0};
+    Timer sut(TIMEOUT, [&] {
+        ++counter;
+        // wait slightly longer then the timeout so that the effect is better measurable
+        std::this_thread::sleep_for(std::chrono::microseconds(TIMEOUT.milliSeconds<int>() * 1100));
+    });
+
+    sut.start(Timer::RunMode::PERIODIC, Timer::CatchUpPolicy::SKIP_TO_NEXT_BEAT);
+
+    std::this_thread::sleep_for(std::chrono::milliseconds(TIMEOUT.milliSeconds<int>() * 100));
+    // every second callback is skipped since the runtime is slightly longer therefore
+    // the counter must be in that range
+    TIMING_TEST_EXPECT_TRUE(50 <= counter && counter <= 70);
```

Comment:
the tests would run faster if e.g. the interval would be 100ms and the callback execution time 150ms. If the time from exiting the first timer invocation to the time from entering the second timer invocation would be less than 1 ms the tests fails. On the second timer invocation a semaphore could be posted to continue a blocking timed wait instead of the thread sleep in the main thread.
The `SoftTimerCallsLessCallbacksThanASAPTimer` test should also not be needed with this approach.

Don't know which approach would be more robust, though

Reply:
At the moment I would leave it as it is since this test is also marked as TIMING_TEST.